### ISSUE TITLE
quantize error

### ISSUE THREAD

**Issue by Hong-333** [YuxuanXie/caffe_mobilessd#2]
i run the ristretto quantize then i got error

![image](https://user-images.githubusercontent.com/40624715/71384941-b0cc9900-2627-11ea-8b02-31469f275305.png)

**Comment by Hong-333** [comment]
@YuxuanXie Can't solve this problem? :(

**Comment by YuxuanXie** [comment]
Oh, sorry. I did not see your comment until now. I think its because your protxt file is not in the right format. Check the output layer of your protxt file please.

**Comment by Hong-333** [comment]
> Oh, sorry. I did not see your comment until now. I think its because your protxt file is not in the right format. Check the output layer of your protxt file please.

https://drive.google.com/file/d/11_gZhajiNEoaCWj4bDG-GB1n0kI2q6K0/view?usp=sharing

this is my prototxt ... is it wrong?

**Comment by YuxuanXie** [comment]
layer {
  name: "detection_out"
  type: "DetectionOutput"
  bottom: "mbox_loc"
  bottom: "mbox_conf_flatten"
  bottom: "mbox_priorbox"
  top: "detection_out"
  include {
    phase: TEST
  }
  detection_output_param {
    num_classes: 4
    share_location: true
    background_label_id: 0
    nms_param {
      nms_threshold: 0.45
      top_k: 400
    }
    code_type: CENTER_SIZE
    keep_top_k: 200
    confidence_threshold: 0.01
  }
}
layer {
  name: "detection_eval"
  type: "DetectionEvaluate"
  bottom: "detection_out"
  bottom: "label"
  top: "detection_eval"
  include {
    phase: TEST
  }
  detection_evaluate_param {
    num_classes: 4
    background_label_id: 0
    overlap_threshold: 0.5
    evaluate_difficult_gt: false
  }
}

Here these two layers just appear in test mode, which is wrong. It should also appear in train mode

**Comment by Hong-333** [comment]
> layer {
> name: "detection_out"
> type: "DetectionOutput"
> bottom: "mbox_loc"
> bottom: "mbox_conf_flatten"
> bottom: "mbox_priorbox"
> top: "detection_out"
> include {
> phase: TEST
> }
> detection_output_param {
> num_classes: 4
> share_location: true
> background_label_id: 0
> nms_param {
> nms_threshold: 0.45
> top_k: 400
> }
> code_type: CENTER_SIZE
> keep_top_k: 200
> confidence_threshold: 0.01
> }
> }
> layer {
> name: "detection_eval"
> type: "DetectionEvaluate"
> bottom: "detection_out"
> bottom: "label"
> top: "detection_eval"
> include {
> phase: TEST
> }
> detection_evaluate_param {
> num_classes: 4
> background_label_id: 0
> overlap_threshold: 0.5
> evaluate_difficult_gt: false
> }
> }
> 
> Here these two layers just appear in test mode, which is wrong. It should also appear in train mode

I removed that part but i got another error 

I0114 08:08:55.524070  3804 net.cpp:830] Ignoring source layer mbox_conf_flatten
I0114 08:08:55.524072  3804 net.cpp:830] Ignoring source layer detection_out
I0114 08:08:55.524210  3804 quantization.cpp:185] Running for 1 iterations.
I0114 08:08:55.524224  3804 blocking_queue.cpp:50] Data layer prefetch queue empty
F0114 08:08:56.795418  3804 quantization.cpp:221] Check failed: result[j]->width() == 5 (8 vs. 5) 
*** Check failure stack trace: ***
    @     0x7fd21df4e0cd  google::LogMessage::Fail()
    @     0x7fd21df4ff33  google::LogMessage::SendToLog()
    @     0x7fd21df4dc28  google::LogMessage::Flush()
    @     0x7fd21df50999  google::LogMessageFatal::~LogMessageFatal()
    @     0x7fd21e3b1944  Quantization::RunForwardBatches()
    @     0x7fd21e3b575c  Quantization::QuantizeNet()
    @     0x5574507d587b  quantize()
    @     0x5574507d4aea  main
    @     0x7fd21d1a7b97  __libc_start_main
    @     0x5574507d53fa  _start
Aborted (core dumped)

**Comment by Hong-333** [comment]
I checked your example so I passed quantize.
I made this prototxt

https://drive.google.com/open?id=19I7SZdpyiRKNEVLW6FGvzXyu9s47N45C

Should I train right away using this prototxt?

**Comment by YuxuanXie** [comment]
Yep,  but this method is designed for the trained model. You should train your model first, and then quantize it by this tool.

**Comment by Hong-333** [comment]
I mean, I have a trained model and I quantized it and I want to finetune that quantization to make it more accurate.  But if I use that as it is and then fine tune, the accuracy does not seem to rise.

**Comment by YuxuanXie** [comment]
Aha, sorry, for int8, It seems like fine tune here doesn't work because of the algorithm we use. It might work for int16. But there is no loss when quantize it into int16. For int8, the loss should be less than 1.5%.

**Comment by Hong-333** [comment]
In my "dynamic_quantized_nobn.prototxt" 
https://drive.google.com/open?id=19I7SZdpyiRKNEVLW6FGvzXyu9s47N45C
There is no TRAIN layer at the output layer. is it work fine tuning without TRAIN layer?

**Comment by YuxuanXie** [comment]
In fact, I did not use fine tune, I am sorry. Theoretically, you should add train layer into the output layer.

**Comment by Hong-333** [comment]
> In fact, I did not use fine tune, I am sorry. Theoretically, you should add train layer into the output layer.

Yes, thank you for your answers so far! 👍
